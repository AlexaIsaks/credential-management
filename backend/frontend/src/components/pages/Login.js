import React from "react";
import { withRouter } from "react-router-dom";

// Login page
class Login extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      username: null,
      password: "",
      formErrors: null,
      error: null
    };
  }

  // Updates state with user's credentials
  handleLoginChange = (event) => {
    const name = event.target.name;
    const value = event.target.value;

    this.setState({
      [name]: value
    });
  }

  // Submits user's login credentials
  handleLoginSubmission = (event) => {
    event.preventDefault();

    const { username, password } = this.state;

    // Fetch arguments
    const parameters = {
      method: "POST",
      headers: {
        "Content-type": "application/json",
      },
      body: JSON.stringify({ username, password })
    };

    fetch("/login", parameters)
      .then((response) => response.json())
      .then(
        (results) => {
          if (results.error) {
            this.setState({
              password: "",
              formErrors: results.error
            });
            // Login successful
          } else {
            this.props.setLogin(true);
            // Redirect user to portal page
            this.props.history.push({
              pathname: "/portal",
              state: results.token
            });
          }
        },
        (error) => {
          this.setState({
            error
          });
        }
      );
  }

  render() {
    let displayLoginContent = null;

    // An error occurred
    if (this.state.error) {
      displayLoginContent = (
        <p className="pt-5 text-center">
          <span className="d-flex justify-content-center align-items-center fs-3 fw-bold text-danger">
            <ion-icon name="sad-outline"></ion-icon>
          </span>
          <span>
            An error has occurred. Please refresh the page or speak to you
            administrator.
          </span>
        </p>
      );
    } else {
      displayLoginContent = (
        <div className="row py-5">
          <div className="col-10 col-md-6 col-lg-4 offset-1 offset-md-3 offset-lg-4 py-4 px-3 border">
            <h1 className="h5 text-center">Credential Management</h1>
            <h2 className="mb-5 h5 text-center">Login</h2>

            {/*Login form */}
            <form onSubmit={this.handleLoginSubmission} className="form">
              <div className="mb-3">
                <label htmlFor="login-username">Username</label>
                <input
                  type="text"
                  name="username"
                  id="login-username"
                  onChange={this.handleLoginChange}
                  className="form-control"
                  required
                />
              </div>

              <div>
                <label htmlFor="login-password">Password</label>
                <input
                  type="password"
                  name="password"
                  value={this.state.password}
                  id="login-password"
                  autoComplete="off"
                  onChange={this.handleLoginChange}
                  className="form-control"
                  required
                />
              </div>

              {/*Login errors*/}
              {this.state.formErrors && (
                <div className="form__error text-danger">
                  {this.state.formErrors}
                </div>
              )}

              <button
                type="submit"
                className="mt-3 btn btn-outline-secondary w-100"
              >
                Submit
              </button>
            </form>
          </div>
        </div>
      );
    }
    return (
      <main className="pt-4 pt-md-5 min-vh-100 container">
        {/*Main content */}
        {displayLoginContent}
      </main>
    );
  }
}

export default withRouter(Login);
