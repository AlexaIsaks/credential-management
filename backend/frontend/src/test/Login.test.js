import React from "react";
import Login from "../components/pages/Login";
import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";

// Snapshot test: Login component
test("Login component renders correctly", () => {
  const tree = renderer
    .create(
      <BrowserRouter>
        <Login />
      </BrowserRouter>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
