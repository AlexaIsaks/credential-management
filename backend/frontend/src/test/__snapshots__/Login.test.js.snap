// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component renders correctly 1`] = `
<main
  className="pt-4 pt-md-5 min-vh-100 container"
>
  <div
    className="row py-5"
  >
    <div
      className="col-10 col-md-6 col-lg-4 offset-1 offset-md-3 offset-lg-4 py-4 px-3 border"
    >
      <h1
        className="h5 text-center"
      >
        Credential Management
      </h1>
      <h2
        className="mb-5 h5 text-center"
      >
        Login
      </h2>
      <form
        className="form"
        onSubmit={[Function]}
      >
        <div
          className="mb-3"
        >
          <label
            htmlFor="login-username"
          >
            Username
          </label>
          <input
            className="form-control"
            id="login-username"
            name="username"
            onChange={[Function]}
            required={true}
            type="text"
          />
        </div>
        <div>
          <label
            htmlFor="login-password"
          >
            Password
          </label>
          <input
            autoComplete="off"
            className="form-control"
            id="login-password"
            name="password"
            onChange={[Function]}
            required={true}
            type="password"
            value=""
          />
        </div>
        <button
          className="mt-3 btn btn-outline-secondary w-100"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>
</main>
`;
